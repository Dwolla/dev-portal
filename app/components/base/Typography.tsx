/** @jsx jsx */
import styled from "@emotion/styled";
import { css, jsx } from "@emotion/core";
import classnames from "../../modules/classnames";
import { POPPINS, ROBOTO, MONACO } from "../typography";
import {
  HEADLINE_TEXT,
  PARAGRAPH_TEXT,
  WHITE_PRIMARY,
  ORANGE_PRIMARY,
  PURPLE_DARK,
  GREY_1,
  GREY_3,
  GREY_4,
  GREY_5,
} from "../colors";

// Heading styles
export const headingStyles = {
  global: css`
    color: ${HEADLINE_TEXT};
    font-family: ${POPPINS};
    font-weight: 300;
    &.dark {
      color: ${WHITE_PRIMARY};
    }
  `,
  h1: css`
    font-size: 42px;
    line-height: 52px;
    margin-top: 26px;
    margin-bottom: 26px;
  `,
  h2: css`
    font-size: 30px;
    line-height: 46px;
    margin-top: 26px;
    margin-bottom: 26px;
  `,
  h3: css`
    font-size: 22px;
    line-height: 33px;
    margin-top: 29px;
    margin-bottom: 29px;
  `,
  h4: css`
    font-size: 18px;
    line-height: 27px;
    margin-top: 36px;
    margin-bottom: 26px;
  `,
  h5: css`
    text-transform: uppercase;
    font-size: 14px;
    font-weight: 500;
    letter-spacing: 1px;
    margin-top: 36px;
    margin-bottom: 26px;
  `,
};

const StyledPreHeader = styled.p`
  color: ${GREY_5};
  text-transform: uppercase;
  font-family: ${POPPINS};
  font-size: 16px;
  font-weight: 500;
  letter-spacing: 1.14px;
  line-height: 21px;
  margin-top: 0;
  margin-bottom: 15px;
  &.dark {
    color: ${GREY_4};
  }
`;

// Paragraph syles
const baseTextStyles = css`
  color: ${PARAGRAPH_TEXT};
  font-family: ${ROBOTO};
  &.dark {
    color: ${GREY_4};
  }
`;

const StyledSubHeader = styled.p`
  ${baseTextStyles}
  font-size: 20px;
  font-weight: 300;
  line-height: 26px;
  margin-top: 26px;
  margin-bottom: 26px;
`;

export const paragraphStyles = css`
  ${baseTextStyles}
  font-size: 16px;
  line-height: 28px;
  margin-top: 26px;
  margin-bottom: 29px;
`;

export const codeStyles = css`
  color: ${PARAGRAPH_TEXT};
  font-family: ${MONACO};
  font-size: 13px;
  line-height: 20px;
  border: 1px solid ${GREY_3};
  border-radius: 5px;
  background-color: ${GREY_1};
  padding: 2px 5px;
  &.dark {
    background-color: ${PURPLE_DARK};
    color: ${GREY_4};
  }
`;

export const linkStyles = css`
  color: ${ORANGE_PRIMARY};
  cursor: pointer;
  text-decoration: none;
  &:hover {
    text-decoration: underline;
  }
`;

// List styles
export const listStyles = css`
  ${baseTextStyles}
  font-size: 16px;
  letter-spacing: 0;
  line-height: 28px;
  margin-top: 26px;
  margin-bottom: 29px;
  > li {
    margin-bottom: 8px;
  }
`;

// Prop types
type Props = {
  children: any;
  isDark?: boolean;
};

// Link Prop types
type LinkProps = {
  text: string;
  href: string;
};

export const PreHeader = ({ children, isDark }: Props) => (
  <StyledPreHeader className={classnames({ dark: isDark })}>
    {children}
  </StyledPreHeader>
);

export const H1 = ({ children, isDark }: Props) => (
  <h1
    css={[headingStyles.global, headingStyles.h1]}
    className={classnames({ dark: isDark })}
  >
    {children}
  </h1>
);

export const H2 = ({ children, isDark }: Props) => (
  <h2
    css={[headingStyles.global, headingStyles.h2]}
    className={classnames({ dark: isDark })}
  >
    {children}
  </h2>
);

export const H3 = ({ children, isDark }: Props) => (
  <h3
    css={[headingStyles.global, headingStyles.h3]}
    className={classnames({ dark: isDark })}
  >
    {children}
  </h3>
);

export const H4 = ({ children, isDark }: Props) => (
  <h4
    css={[headingStyles.global, headingStyles.h4]}
    className={classnames({ dark: isDark })}
  >
    {children}
  </h4>
);

export const H5 = ({ children, isDark }: Props) => (
  <h5
    css={[headingStyles.global, headingStyles.h5]}
    className={classnames({ dark: isDark })}
  >
    {children}
  </h5>
);

export const SubHeader = ({ children, isDark }: Props) => (
  <StyledSubHeader className={classnames({ dark: isDark })}>
    {children}
  </StyledSubHeader>
);

export const Paragraph = ({ children, isDark }: Props) => (
  <p css={paragraphStyles} className={classnames({ dark: isDark })}>
    {children}
  </p>
);

export const UnorderedList = ({ children, isDark }: Props) => (
  <ul
    css={css`
      ${listStyles}
    `}
    className={classnames({ dark: isDark })}
  >
    {children.map((item) => (
      <li key={item.id}>{item.text}</li>
    ))}
  </ul>
);

export const OrderedList = ({ children, isDark }: Props) => (
  <ol
    css={css`
      ${listStyles}
    `}
    className={classnames({ dark: isDark })}
  >
    {children.map((item) => (
      <li key={item.id}>{item.text}</li>
    ))}
  </ol>
);

export const BlockQuote = ({ children, isDark }: Props) => (
  <blockquote
    css={css`
      ${baseTextStyles}
    `}
    className={classnames({ dark: isDark })}
  >
    {children}
  </blockquote>
);

export const InlineCode = ({ children, isDark }: Props) => (
  <code css={codeStyles} className={classnames({ dark: isDark })}>
    {children}
  </code>
);

export const Link = ({ text, href }: LinkProps) => (
  <a css={linkStyles} href={href}>
    {text}
  </a>
);
