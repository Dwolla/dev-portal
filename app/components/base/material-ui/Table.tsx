import React, { Fragment, ReactElement, useState } from "react";
import {
  IconButton,
  Paper,
  styled,
  Table as MuiTable,
  TableBody,
  TableCell,
  TableContainer as MuiTableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@mui/material";
import { ExpandLessOutlined, ExpandMoreOutlined } from "@mui/icons-material";
import { v4 as uuidv4 } from "uuid";
import { slugify } from "../../../modules/helpers";
import { GREY_1, GREY_3_44, PURPLE_012 } from "../../colors";

const ExpandableTableCell = styled(TableCell, { name: "ExpandableTableCell" })(
  () => ({
    width: "3.5rem",
  })
);

const NestedTableContainer = styled(MuiTableContainer, {
  name: "NestedTableContainer",
})(() => ({
  padding: "1rem 2rem",
}));

const NestedTableWrapper = styled(Paper, { name: "NestedTable" })(() => ({
  background: `${GREY_1}`,
  boxShadow: `0px 0px 0px 1px ${PURPLE_012};`,
  padding: "1rem",
}));

const TitleTypography = styled(Typography, { name: "TitleTypography" })(() => ({
  background: `${GREY_3_44}`,
  borderRadius: "0.25rem",
  padding: "0.5rem",
}));

export interface TableProps {
  table: TableContents | NestedTableContents;
}

export interface TableContents {
  headers: string[];
  rows: ContentRow[];
}

export interface NestedTableContents extends TableContents {
  title: string;

  /**
   * This property is optional since it's autopopulated when the component renders.
   *
   * A custom slugged title can be added here; however, it's best to let the component
   * generate it since it must be unique across all nested tables within the current page.
   */
  sluggedTitle?: string;
}

type Content = ReactElement | string;
type ContentRow = Content[] | NestedTableContentRow;
type NestedTableContentRow = [Content[], NestedTableContents];
type SluggedNestedTableContentRow = [
  Content[],
  NestedTableContents & Required<Pick<NestedTableContents, "sluggedTitle">>
];

const isElementOrString = (obj: any): obj is ReactElement | string =>
  React.isValidElement(obj) || typeof obj === "string";

const isNestedTableContents = (obj: any): obj is NestedTableContents =>
  typeof obj?.title !== "undefined";

const isNestedTableRow = (obj: any): obj is NestedTableContentRow =>
  obj.length === 2 &&
  Array.isArray(obj[0]) &&
  obj[0].every(isElementOrString) &&
  typeof obj[1]?.title !== "undefined";

const isSluggedNestedTableRow = (
  obj: any
): obj is SluggedNestedTableContentRow =>
  isNestedTableRow(obj) && typeof obj[1]?.sluggedTitle !== "undefined";

const isTableExpandable = (table: TableContents): boolean =>
  table.rows.some(isNestedTableRow);

const addUniqueIdToNestedTabled = (table: TableContents): TableContents => ({
  ...table,
  rows: table.rows.map((item): ContentRow => {
    if (isNestedTableRow(item) && typeof item[1].sluggedTitle === "undefined") {
      // eslint-disable-next-line no-param-reassign
      item[1].sluggedTitle = `${slugify(item[1].title)}-${uuidv4()}`;
    }
    return item;
  }),
});

export default function Table({ table }: TableProps): ReactElement {
  const sluggedTable = addUniqueIdToNestedTabled(table);

  const TableContainer = isNestedTableContents(sluggedTable)
    ? NestedTableContainer
    : MuiTableContainer;
  const TableWrapper = isNestedTableContents(sluggedTable)
    ? NestedTableWrapper
    : Fragment;

  const [expanded, setExpanded] = useState<string[]>([]);
  const addExpanded = (slug: string): string[] => [...expanded, slug];
  const removeExpanded = (slug: string): string[] =>
    expanded.filter((value) => value !== slug);

  const handleExpandClicked = (row: SluggedNestedTableContentRow) => (): void =>
    !expanded.includes(row[1].sluggedTitle)
      ? setExpanded([...addExpanded(row[1].sluggedTitle)])
      : setExpanded([...removeExpanded(row[1].sluggedTitle)]);

  const getCellElements = (row: ContentRow): ReactElement[] => {
    const cells = isNestedTableRow(row) ? row[0] : row;
    const borderBottom =
      isSluggedNestedTableRow(row) && expanded.includes(row[1].sluggedTitle)
        ? 0
        : null;
    const mappedCells = cells.map((cell, cellIndex) => (
      <TableCell key={`cell-${cellIndex}`} sx={{ borderBottom }}>
        {cell}
      </TableCell>
    ));

    if (isSluggedNestedTableRow(row)) {
      const ExpandIcon = expanded.includes(row[1].sluggedTitle)
        ? ExpandLessOutlined
        : ExpandMoreOutlined;

      mappedCells.push(
        <TableCell sx={{ borderBottom }}>
          <IconButton onClick={handleExpandClicked(row)} size="small">
            <ExpandIcon />
          </IconButton>
        </TableCell>
      );
    } else if (!isNestedTableContents(sluggedTable)) {
      mappedCells.push(<TableCell></TableCell>);
    }
    return mappedCells;
  };

  return (
    <TableContainer>
      <TableWrapper>
        {isNestedTableContents(sluggedTable) && (
          <TitleTypography>{sluggedTable.title}</TitleTypography>
        )}

        <MuiTable>
          <TableHead>
            <TableRow>
              {sluggedTable.headers.map((item, itemIndex) => (
                <TableCell key={`item-${itemIndex}`}>{item}</TableCell>
              ))}

              {isTableExpandable(sluggedTable) && <ExpandableTableCell />}
            </TableRow>
          </TableHead>

          <TableBody>
            {sluggedTable.rows.map((row, rowIndex) => (
              <React.Fragment key={`row-${rowIndex}`}>
                <TableRow>{getCellElements(row)}</TableRow>

                {isSluggedNestedTableRow(row) &&
                  expanded.includes(row[1].sluggedTitle) && (
                    <TableRow>
                      <TableCell colSpan={99} sx={{ padding: 0 }}>
                        <Table table={row[1]} />
                      </TableCell>
                    </TableRow>
                  )}
              </React.Fragment>
            ))}
          </TableBody>
        </MuiTable>
      </TableWrapper>
    </TableContainer>
  );
}
