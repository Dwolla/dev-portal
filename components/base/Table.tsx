import React from "react";
import styled from "@emotion/styled";
import { ROBOTO } from "../typography";
import { HEADLINE_TEXT, PARAGRAPH_TEXT, GREY_1, GREY_2 } from "../colors";
import { breakUp } from "../breakpoints";

// Styles for Table
const StyledTable = styled.div`
  overflow-x: scroll;
  white-space: nowrap;
  /* This allows text within cells to wrap around on larger screens */
  @media (${breakUp("lg")}) {
    white-space: unset;
  }

  table {
    /* Separate borders to selectively apply border radius */
    width: 100%;
    border-collapse: separate;
    border-spacing: 0;
    text-align: left;
    font-family: ${ROBOTO};
  }
  th {
    color: ${HEADLINE_TEXT};
    font-size: 13px;
    font-weight: 500;
    line-height: 17px;
    padding: 11px 21px;
    text-align: left;
  }
  td {
    color: ${PARAGRAPH_TEXT};
    font-size: 14px;
    line-height: 20px;
    padding: 17px 21px;
    border-top: 1px solid ${GREY_2};
    border-bottom: 1px solid ${GREY_2};
    :nth-of-type(1) {
      border-left: 1px solid ${GREY_2};
    }
    :nth-last-of-type(1) {
      border-right: 1px solid ${GREY_2};
      /* This ensures the last column width doesn't shrink to a point where the text wraps around each word */
      @media (${breakUp("lg")}) {
        min-width: 20ch;
      }
    }
  }
  /* Applying background-color to alternate rows  */
  tr:nth-of-type(even) {
    background-color: ${GREY_1};
  }

  /* td border-radius */
  tbody {
    tr {
      &:nth-of-type(1) {
        td {
          &:nth-of-type(1) {
            border-top-left-radius: 5px;
          }
          &:last-of-type {
            border-top-right-radius: 5px;
          }
        }
      }
      &:last-of-type {
        td {
          &:nth-of-type(1) {
            border-bottom-left-radius: 5px;
          }
          &:last-of-type {
            border-bottom-right-radius: 5px;
          }
        }
      }
    }
  }
`;

// Prop types
type Props = {
  children: JSX.Element;
};

// Table component
export default function Table({ children }: Props) {
  return <StyledTable>{children}</StyledTable>;
}
