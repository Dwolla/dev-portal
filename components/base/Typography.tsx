/** @jsx jsx */
import styled from "@emotion/styled";
import { css, jsx } from "@emotion/core";
import classnames from "../../modules/classnames";
import { POPPINS, ROBOTO, MONACO } from "../typography";
import {
  HEADLINE_TEXT,
  PARAGRAPH_TEXT,
  WHITE_PRIMARY,
  ORANGE_PRIMARY,
  PURPLE_DARK,
  GREY_1,
  GREY_3,
  GREY_4,
  GREY_5,
} from "../colors";

// Heading styles
const headingStyles = css`
  color: ${HEADLINE_TEXT};
  font-family: ${POPPINS};
  font-weight: 300;
  &.dark {
    color: ${WHITE_PRIMARY};
  }
`;

const StyledPreHeader = styled.span`
  color: ${GREY_5};
  text-transform: uppercase;
  font-family: ${POPPINS};
  font-size: 16px;
  font-weight: 500;
  letter-spacing: 1.14px;
  line-height: 21px;
  &.dark {
    color: ${GREY_4};
  }
`;

const StyledH1 = styled.h1`
  ${headingStyles}
  font-size: 42px;
  line-height: 52px;
`;
const StyledH2 = styled.h2`
  ${headingStyles}
  font-size: 30px;
  line-height: 46px;
`;
const StyledH3 = styled.h3`
  ${headingStyles}
  font-size: 22px;
  line-height: 33px;
`;
const StyledH4 = styled.h4`
  ${headingStyles}
  font-size: 18px;
  line-height: 27px;
`;
const StyledH5 = styled.h5`
  ${headingStyles}
  text-transform: uppercase;
  font-size: 14px;
  font-weight: 500;
  letter-spacing: 1px;
`;

// Paragraph syles
const ParagraphStyles = css`
  color: ${PARAGRAPH_TEXT};
  font-family: ${ROBOTO};
  &.dark {
    color: ${GREY_4};
  }
`;

const StyledSubHeader = styled.span`
  ${ParagraphStyles}
  font-size: 20px;
  font-weight: 300;
  line-height: 26px;
`;

const StyledParagraph = styled.p`
  ${ParagraphStyles}
  font-size: 16px;
  line-height: 28px;
`;

const StyledInlineCode = styled.code`
  color: ${PARAGRAPH_TEXT};
  font-family: ${MONACO};
  font-size: 13px;
  line-height: 20px;
  border: 1px solid ${GREY_3};
  border-radius: 5px;
  background-color: ${GREY_1};
  padding: 2px 5px;
  &.dark {
    background-color: ${PURPLE_DARK};
    color: ${GREY_4};
  }
`;

const StyledLink = styled.a`
  color: ${ORANGE_PRIMARY};
  text-decoration: none;
  &:hover {
    text-decoration: underline;
  }
`;

// List styles
const ListStyles = css`
  ${ParagraphStyles}
  font-size: 16px;
  letter-spacing: 0;
  line-height: 28px;
`;

// Prop types
type Props = {
  children: any;
  isDark?: boolean;
};

// Link Prop types
type LinkProps = {
  text: string;
  href: string;
};

export const PreHeader = ({ children, isDark }: Props) => (
  <StyledPreHeader className={classnames({ dark: isDark })}>
    {children}
  </StyledPreHeader>
);

export const H1 = ({ children, isDark }: Props) => (
  <StyledH1 className={classnames({ dark: isDark })}>{children}</StyledH1>
);

export const H2 = ({ children, isDark }: Props) => (
  <StyledH2 className={classnames({ dark: isDark })}>{children}</StyledH2>
);

export const H3 = ({ children, isDark }: Props) => (
  <StyledH3 className={classnames({ dark: isDark })}>{children}</StyledH3>
);

export const H4 = ({ children, isDark }: Props) => (
  <StyledH4 className={classnames({ dark: isDark })}>{children}</StyledH4>
);

export const H5 = ({ children, isDark }: Props) => (
  <StyledH5 className={classnames({ dark: isDark })}>{children}</StyledH5>
);

export const SubHeader = ({ children, isDark }: Props) => (
  <StyledSubHeader className={classnames({ dark: isDark })}>
    {children}
  </StyledSubHeader>
);

export const Paragraph = ({ children, isDark }: Props) => (
  <StyledParagraph className={classnames({ dark: isDark })}>
    {children}
  </StyledParagraph>
);

export const UnorderedList = ({ children, isDark }: Props) => (
  <ul
    css={css`
      ${ListStyles}
    `}
    className={classnames({ dark: isDark })}
  >
    {children.map((item) => (
      <li key={item.id}>{item.text}</li>
    ))}
  </ul>
);

export const OrderedList = ({ children, isDark }: Props) => (
  <ol
    css={css`
      ${ListStyles}
    `}
    className={classnames({ dark: isDark })}
  >
    {children.map((item) => (
      <li key={item.id}>{item.text}</li>
    ))}
  </ol>
);

export const BlockQuote = ({ children, isDark }: Props) => (
  <blockquote
    css={css`
      ${ParagraphStyles}
    `}
    className={classnames({ dark: isDark })}
  >
    {children}
  </blockquote>
);

export const InlineCode = ({ children, isDark }: Props) => (
  <StyledInlineCode className={classnames({ dark: isDark })}>
    {children}
  </StyledInlineCode>
);

export const Link = ({ text, href }: LinkProps) => (
  <StyledLink href={href}>{text}</StyledLink>
);
