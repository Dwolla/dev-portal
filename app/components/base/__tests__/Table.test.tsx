import renderer from "react-test-renderer";
import Table from "../Table";
import { InlineCode } from "../Typography";

const tableData = (
  <>
    <tr>
      <th>Parameter</th>
      <th>Required?</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
    <tr>
      <td>firstName</td>
      <td>yes</td>
      <td>string</td>
      <td>Customer&apos;s first name</td>
    </tr>
    <tr>
      <td>lastName</td>
      <td>yes</td>
      <td>string</td>
      <td>Customer&apos;s last name</td>
    </tr>
    <tr>
      <td>email</td>
      <td>yes</td>
      <td>string</td>
      <td>Customer&apos;s last name</td>
    </tr>
    <tr>
      <td>type</td>
      <td>yes</td>
      <td>string</td>
      <td>
        Value of <InlineCode>receive-only</InlineCode>
      </td>
    </tr>
    <tr>
      <td>businessName</td>
      <td>yes</td>
      <td>string</td>
      <td>Customer&apos;s registered business name</td>
    </tr>
    <tr>
      <td>ipAddress</td>
      <td>yes</td>
      <td>string</td>
      <td>Customer&apos;s IP address</td>
    </tr>
  </>
);

test("Table", () => {
  const tree = renderer.create(<Table>{tableData}</Table>).toJSON();

  expect(tree).toMatchSnapshot();
});
