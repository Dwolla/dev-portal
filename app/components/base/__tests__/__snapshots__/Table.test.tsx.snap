// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table 1`] = `
<table
  className="css-8bi4x2-StyledTable e1rel8440"
>
  <tr>
    <th>
      Parameter
    </th>
    <th>
      Required?
    </th>
    <th>
      Type
    </th>
    <th>
      Description
    </th>
  </tr>
  <tr>
    <td>
      firstName
    </td>
    <td>
      yes
    </td>
    <td>
      string
    </td>
    <td>
      Customer's first name
    </td>
  </tr>
  <tr>
    <td>
      lastName
    </td>
    <td>
      yes
    </td>
    <td>
      string
    </td>
    <td>
      Customer's last name
    </td>
  </tr>
  <tr>
    <td>
      email
    </td>
    <td>
      yes
    </td>
    <td>
      string
    </td>
    <td>
      Customer's last name
    </td>
  </tr>
  <tr>
    <td>
      type
    </td>
    <td>
      yes
    </td>
    <td>
      string
    </td>
    <td>
      Value of 
      <code
        className=" css-wuyv6i-codeStyles-codeStyles"
      >
        receive-only
      </code>
    </td>
  </tr>
  <tr>
    <td>
      businessName
    </td>
    <td>
      yes
    </td>
    <td>
      string
    </td>
    <td>
      Customer's registered business name
    </td>
  </tr>
  <tr>
    <td>
      ipAddress
    </td>
    <td>
      yes
    </td>
    <td>
      string
    </td>
    <td>
      Customer's IP address
    </td>
  </tr>
</table>
`;
